jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/RetrieveRegistryCredentialsStep.java

```diff
@@ -87,20 +98,30 @@ static RetrieveRegistryCredentialsStep forTargetImage(
   @Nullable
   public Credential call() throws CredentialRetrievalException {
     String description = makeDescription(registry);
-    eventDispatcher.dispatch(LogEvent.lifecycle(description + "..."));
 
-    try (TimerEventDispatcher ignored = new TimerEventDispatcher(eventDispatcher, description)) {
+    buildConfiguration.getEventDispatcher().dispatch(LogEvent.lifecycle(description + "..."));
+    Allocation progressAllocation =
+        parentProgressAllocation.newChild("Retrieve registry credentials for " + registry, 1);
+    buildConfiguration.getEventDispatcher().dispatch(new ProgressEvent(progressAllocation, 0));
+
+    try (TimerEventDispatcher ignored =
+        new TimerEventDispatcher(buildConfiguration.getEventDispatcher(), description)) {
       for (CredentialRetriever credentialRetriever : credentialRetrievers) {
         Optional<Credential> optionalCredential = credentialRetriever.retrieve();
         if (optionalCredential.isPresent()) {
+          buildConfiguration
+              .getEventDispatcher()
+              .dispatch(new ProgressEvent(progressAllocation, 1));
```

Comment:
Another way is to put this is to add a `finally` block to the existing `try`.

If there is no existing `try`, you can add one, but another way would be to wrap it in another method, like

```java
Object call() {
  try {
   send 0-unit event
   return realCall();
  } finally {
    send completion event
  }
}

Object realCall() {
  // the original method
}
```

But sometimes, just putting the final event in two places can be simpler, and I have no object to it.

Reply:
Yea, this will be completely refactored to not include so much duplication after `ProgressEventDispatcher` is added.